odl/test/discr/discr_utils_test.py

```diff
@@ -0,0 +1,265 @@
+    assert val == 3.0
+    # Input array, with and without output array
+    pts = np.array([[0.3, 0.6],
+                    [1.0, 1.0]])
+    true_arr = [3, 7]
+    assert all_equal(interpolator(pts.T), true_arr)
+    out = np.empty(2, dtype='float64')
+    interpolator(pts.T, out=out)
+    assert all_equal(out, true_arr)
+    # Input meshgrid, with and without output array
+    mg = sparse_meshgrid([0.3, 1.0], [0.4, 1.0])
+    # Indices: (1, 3) x (0, 1)
+    true_mg = [[2, 3],
+               [6, 7]]
+    assert all_equal(interpolator(mg), true_mg)
+    out = np.empty((2, 2), dtype='float64')
+    interpolator(mg, out=out)
+    assert all_equal(out, true_mg)
+
+
+def test_nearest_interpolation_2d_string():
+    """Test nearest neighbor interpolation in 2d with string values."""
+    coord_vecs = [[0.125, 0.375, 0.625, 0.875], [0.25, 0.75]]
+    f = np.array(list('mystring'), dtype='U1').reshape((4, 2))
+    interpolator = nearest_interpolator(f, coord_vecs)
+
+    # Evaluate at single point
+    val = interpolator([0.3, 0.6])  # closest to index (1, 1) -> 3
+    assert val == u't'
+    # Input array, with and without output array
+    pts = np.array([[0.3, 0.6],
+                    [1.0, 1.0]])
+    true_arr = np.array(['t', 'g'], dtype='U1')
+    assert all_equal(interpolator(pts.T), true_arr)
+    out = np.empty(2, dtype='U1')
+    interpolator(pts.T, out=out)
+    assert all_equal(out, true_arr)
+    # Input meshgrid, with and without output array
+    mg = sparse_meshgrid([0.3, 1.0], [0.4, 1.0])
+    # Indices: (1, 3) x (0, 1)
+    true_mg = np.array([['s', 't'],
+                        ['n', 'g']], dtype='U1')
+    assert all_equal(interpolator(mg), true_mg)
+    out = np.empty((2, 2), dtype='U1')
+    interpolator(mg, out=out)
+    assert all_equal(out, true_mg)
+
+
+def test_linear_interpolation_1d():
+    """Test linear interpolation in 1d."""
+    coord_vecs = [[0.1, 0.3, 0.5, 0.7, 0.9]]
+    f = np.array([1, 2, 3, 4, 5], dtype="float64")
+    interpolator = linear_interpolator(f, coord_vecs)
+
+    # Evaluate at single point
+    val = interpolator(0.35)
+    true_val = 0.75 * 2 + 0.25 * 3
+    assert val == pytest.approx(true_val)
+
+    # Input array, with and without output array
+    pts = np.array([0.4, 0.0, 0.65, 0.95])
+    true_arr = [2.5, 0.5, 3.75, 3.75]
+    assert all_almost_equal(interpolator(pts), true_arr)
+
+
+def test_linear_interpolation_2d():
+    """Test linear interpolation in 2d."""
+    coord_vecs = [[0.125, 0.375, 0.625, 0.875], [0.25, 0.75]]
+    f = np.arange(1, 9, dtype='float64').reshape((4, 2))
+    interpolator = linear_interpolator(f, coord_vecs)
+
+    # Evaluate at single point
+    val = interpolator([0.3, 0.6])
+    l1 = (0.3 - 0.125) / (0.375 - 0.125)
+    l2 = (0.6 - 0.25) / (0.75 - 0.25)
+    true_val = (
+        (1 - l1) * (1 - l2) * f[0, 0]
+        + (1 - l1) * l2 * f[0, 1]
+        + l1 * (1 - l2) * f[1, 0]
+        + l1 * l2 * f[1, 1]
+    )
+    assert val == pytest.approx(true_val)
```

Comment:
Confession: I only checked this one carefully - the checks below in `test_linear_interpolation_2d` I have not done explicitly. But they have the same structure, and therefore seems to be correct.

Reply:
> Confession

:rofl: 

I can relate. But these tests are really the same as the old ones in `discr_mappings_test`, so I'm not too worried about correctness.